x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,168 @@
+                if (geometry instanceof Point point) {
+                    return new PushableGeoDistance(fieldAttribute, order, point);
+                }
+            } else if (distance.right() instanceof FieldAttribute fieldAttribute && distance.left().foldable()) {
+                Geometry geometry = SpatialRelatesUtils.makeGeometryFromLiteral(distance.left());
+                if (geometry instanceof Point point) {
+                    return new PushableGeoDistance(fieldAttribute, order, point);
+                }
             }
+            return null;
+        }
+    }
+
+    record PushableCompoundExec(EvalExec evalExec, EsQueryExec queryExec, List<EsQueryExec.Sort> pushableSorts) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            // We need to keep the EVAL in place because the coordinator will have its own TopNExec so we need to keep the distance
+            return evalExec.replaceChild(queryExec.withSorts(pushableSorts).withLimit(topNExec.limit()));
         }
-        return plan;
     }
 
-    private static boolean canPushSorts(PhysicalPlan plan) {
-        if (plan instanceof EsQueryExec queryExec) {
-            return queryExec.canPushSorts();
+    private static Pushable evaluatePushable(TopNExec topNExec, Predicate<FieldAttribute> hasIdenticalDelegate) {
+        PhysicalPlan child = topNExec.child();
+        if (child instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // With the simplest case of `FROM index | SORT ...` we only allow pushing down if the sort is on a field
+            return new PushableQueryExec(queryExec);
         }
-        if (plan instanceof ExchangeExec exchangeExec && exchangeExec.child() instanceof EsQueryExec queryExec) {
-            return queryExec.canPushSorts();
+        if (child instanceof ExchangeExec exchangeExec
+            && exchangeExec.child() instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // When we have an exchange between the FROM and the SORT, we also only allow pushing down if the sort is on a field
+            return new PushableExchangeExec(exchangeExec, queryExec);
+        }
+        if (child instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec && queryExec.canPushSorts()) {
+            // When we have an EVAL between the FROM and the SORT, we consider pushing down if the sort is on a field and/or
+            // a distance function defined in the EVAL. We also move the EVAL to after the SORT.
+            List<Order> orders = topNExec.order();
+            List<Alias> fields = evalExec.fields();
+            LinkedHashMap<NameId, StDistance> refs = fields.stream().reduce(new LinkedHashMap<>(), (m, a) -> {
```

Comment:
I am wondering if you considered `AttributeMap` usage here instead of this.

Reply:
I'm looking into that, but it does not seem like a correct match for my needs. I need to lookup from ReferenceAttributes found later to the StDistance function within the Alias that the ReferenceAttribute points to. However there is no attribute available at map creation, only the Alias.id() which is later referred to. I could imagine doing a two-pass algorithm over the orders and the aliases to try build an AttributeMap, but that would require this intermediate map anyway. The current approach seems to be a much simpler solution.